src/week2.hs

```diff
@@ -0,0 +1,202 @@
+{-# OPTIONS_GHC -Wall #-}
+--module week2 where
+
+-- Mastermind -----------------------------------------
+
+-- A peg can be one of six colors
+data Peg = Red | Green | Blue | Yellow | Orange | Purple
+         deriving (Show, Eq, Ord, Enum)
+
+-- A code is defined to simply be a list of Pegs
+type Code = [Peg]
+
+-- A move is constructed using a Code and two integers; the number of
+-- exact matches and the number of regular matches
+data Move = Move Code Int Int
+          deriving (Show, Eq)
+
+-- List containing all of the different Pegs
+colors :: [Peg]
+colors = [Red, Green, Blue, Yellow, Orange, Purple]
+
+-- Exercise 1 -----------------------------------------
+
+-- Get the number of exact matches between the actual code and the guess
+exactMatches :: Code -> Code -> Int
+
+exactMatches [] _ = 0
+exactMatches _ [] = 0
+exactMatches (x:xs) (x':xs') 
+    |x == x' = 1 + (exactMatches xs xs')
+    |otherwise = (exactMatches xs xs')
+
+-- or, using zipWith (Dmitry's suggestion)
+sumIfEqual :: Peg -> Peg -> Int
+sumIfEqual x y = if (x == y) then 1 else 0
+
+exactMatches1 :: Code -> Code -> Int
+exactMatches1 x y = sum(zipWith (sumIfEqual) x y)
+
+-- or, even more concisely (though I'm not sure if this is better)
+exactMatches2 :: Code -> Code -> Int
+exactMatches2 x y = sum(zipWith (\a b -> if (a == b) then 1 else 0) x y)
+
+-- Exercise 2 -----------------------------------------
+-- Write a function that returns the number of total matches between the secret code and the guess
+
+-- helper
+sumLists :: [Int] -> [Int] -> [Int]
+sumLists [] _ = []
+sumLists _ [] = []
+sumLists (x:xs) (x':xs') = (x + x'):(sumLists xs xs')
+
+--or, with implicit recursion (Jan's suggestion):
+sumLists1 :: [Int] -> [Int] -> [Int]
+sumLists1 a b = zipWith (+) a b
+
+------ function that returns a list with color counts
+countColors :: Code -> [Int]
+countColors [] = [0,0,0,0,0,0]
+countColors (x:xs)
+    | x == Red    = sumLists [1,0,0,0,0,0] (countColors xs)
+    | x == Green  = sumLists [0,1,0,0,0,0] (countColors xs)
+    | x == Blue   = sumLists [0,0,1,0,0,0] (countColors xs)
+    | x == Yellow = sumLists [0,0,0,1,0,0] (countColors xs)
+    | x == Orange = sumLists [0,0,0,0,1,0] (countColors xs)
+    | otherwise   = sumLists [0,0,0,0,0,1] (countColors xs)
+
+-- or, more concisely and without hardcoding:
+pegCount :: Int
+pegCount = length [toEnum 0::Peg ..] -- total number of pegs
+
+-- encode (Dmitry's suggestion) returns an int list of length pegCount that is filled with 0s, except at the index corresponding to the input peg, where the value will be 1
+encode :: Peg -> [Int]
+encode p = (replicate k 0) ++ [1] ++ (replicate (pegCount - k - 1) 0)
+                where k = fromEnum p
+
+countColors1 :: Code -> [Int]
+countColors1 c = case c of   []     -> replicate pegCount 0
+                             (x:xs) -> sumLists1 (encode x) (countColors xs)
```

Comment:
awesome!! 😄

Reply:
check out the [LambdaCase](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/lambda_case.html) language extension---it's slightly shorten this definition to:
```haskell
countColors1 = \case
  []       -> replicate pegCount 0                             
  (x:xs) -> sumLists1 (encode x) (countColors xs)
```
it'll also be one of the most gentle language extensions---which we make major use of!---to start off with